File: backend/routes/attendance_routes.py
from flask import Blueprint, request, jsonify, Response
from datetime import datetime
import geo_utils
import csv
import io
from db import attendance_collection, mongo  # Import mongo to access office_config

attendance_bp = Blueprint("attendance", __name__)

# -------------------- PUNCH IN (DYNAMIC) --------------------
@attendance_bp.route("/punch-in", methods=["POST"])
def punch_in():
    data = request.get_json()

    employee_id = data.get("employeeId")
    location = data.get("location")
    wifi_bssid = data.get("wifiBSSID")

    print(f"--- New Punch-In Request ---")
    print(f"Employee: {employee_id}")
    print(f"WiFi BSSID: {wifi_bssid}")
    print(f"Location: {location}")

    if not employee_id or not location or not wifi_bssid:
        return jsonify({
            "message": "employeeId, location and wifiBSSID required",
            "status": "INVALID_REQUEST"
        }), 400

    office_config = mongo.db.office_config.find_one({"branch_name": "Main Office"})
    if not office_config:
        print("❌ ERROR: 'Main Office' not found!")
        return jsonify({
            "message": "Office not configured",
            "status": "OFFICE_NOT_CONFIGURED"
        }), 500

    OFFICE_LAT = office_config['location']['latitude']
    OFFICE_LON = office_config['location']['longitude']
    GEOFENCE_RADIUS = office_config['location']['allowed_radius_meters']
    ALLOWED_BSSIDS = [b.lower() for b in office_config['wifi']['allowed_bssids']]

    # 🔒 Already punched in
    existing = attendance_collection.find_one({
        "employeeId": employee_id,
        "punchOutTime": {"$exists": False}
    })

    if existing:
        print(f"❌ REJECTED: {employee_id} already punched in.")
        return jsonify({
            "message": "Already punched in",
            "status": "PUNCHED_IN"
        }), 409   # ✅ Conflict (better than 400)

    user_lat = location.get("latitude")
    user_lon = location.get("longitude")

    # 🌍 Geofence check
    distance = geo_utils.haversine(OFFICE_LAT, OFFICE_LON, user_lat, user_lon)
    if distance > GEOFENCE_RADIUS:
        print(f"❌ GEOFENCE FAILURE: {round(distance, 2)}m")
        return jsonify({
            "message": "Outside office geofence",
            "status": "GEOFENCE_FAILED",
            "distance_meters": round(distance, 2),
            "allowed_radius": GEOFENCE_RADIUS
        }), 403

    # 📶 WiFi check
    if wifi_bssid.lower() not in ALLOWED_BSSIDS:
        print(f"❌ WIFI FAILURE: {wifi_bssid}")
        return jsonify({
            "message": "Invalid office WiFi",
            "status": "WIFI_FAILED",
            "your_bssid": wifi_bssid
        }), 403

    # ✅ Insert record
    record = {
        "employeeId": employee_id,
        "punchInTime": datetime.utcnow(),
        "location": location,
        "wifiBSSID": wifi_bssid,
        "gpsValid": True,
        "wifiValid": True,
        "status": "PUNCHED_IN",
        "distance_from_office": distance
    }

    attendance_collection.insert_one(record)

    print(f"✅ SUCCESS: {employee_id} punched in.")

    return jsonify({
        "message": "Punch-in successful",
        "status": "PUNCHED_IN"
    }), 200


# -------------------- PUNCH OUT (UNCHANGED) --------------------
@attendance_bp.route("/punch-out", methods=["POST"])
def punch_out():
    data = request.get_json()
    employee_id = data.get("employeeId")

    if not employee_id:
        return jsonify({
            "message": "employeeId is required",
            "status": "INVALID_REQUEST"
        }), 400

    record = attendance_collection.find_one({
        "employeeId": employee_id,
        "punchOutTime": {"$exists": False}
    })

    if not record:
        return jsonify({
            "message": "No active punch-in found",
            "status": "ALREADY_PUNCHED_OUT"
        }), 409   # ✅ Correct semantic code

    attendance_collection.update_one(
        {"_id": record["_id"]},
        {
            "$set": {
                "punchOutTime": datetime.utcnow(),
                "status": "PUNCHED_OUT"
            }
        }
    )

    return jsonify({
        "message": "Punch-out successful",
        "status": "PUNCHED_OUT"
    }), 200


# -------------------- SUMMARY (UNCHANGED) --------------------
@attendance_bp.route("/summary", methods=["GET"])
def attendance_summary():
    employee_id = request.args.get("employeeId")

    if not employee_id:
        return jsonify({"error": "employeeId is required"}), 400

    records = list(
        attendance_collection.find(
            {"employeeId": employee_id},
            {"_id": 0}
        ).sort("punchInTime", -1)  # 🔥 FIX HERE
    )

    total_seconds = 0
    summary = []

    for record in records:
        punch_in = record.get("punchInTime")
        punch_out = record.get("punchOutTime")
        working_hours = None

        if punch_in and punch_out:
            diff = punch_out - punch_in
            total_seconds += diff.total_seconds()
            working_hours = round(diff.total_seconds() / 3600, 2)

        summary.append({
            "employeeId": record.get("employeeId"),
            "punchInTime": punch_in,
            "punchOutTime": punch_out,
            "status": record.get("status"),
            "workingHours": working_hours
        })

    total_hours = round(total_seconds / 3600, 2)

    return jsonify({
        "totalWorkingHours": total_hours,
        "records": summary
    }), 200

# -------------------- EXPORT (UNCHANGED) --------------------
@attendance_bp.route("/export/monthly", methods=["GET"])
def export_monthly_attendance():
    employee_id = request.args.get("employeeId", "").strip()
    month = request.args.get("month", "").strip()

    if not employee_id or not month:
        return jsonify({"error": "employeeId and month are required"}), 400

    try:
        year, month_num = map(int, month.split("-"))
    except ValueError:
        return jsonify({"error": "Invalid month format. Use YYYY-MM"}), 400

    if month_num == 12:
        start_date = datetime(year, 12, 1)
        end_date = datetime(year + 1, 1, 1)
    else:
        start_date = datetime(year, month_num, 1)
        end_date = datetime(year, month_num + 1, 1)

    records = list(
        attendance_collection.find({
            "employeeId": employee_id,
            "punchInTime": {"$gte": start_date, "$lt": end_date}
        })
    )

    if not records:
        return jsonify({"error": "No attendance data found"}), 404

    output = io.StringIO()
    writer = csv.writer(output)
    writer.writerow(["Date", "Punch In", "Punch Out", "Working Hours"])

    total_seconds = 0

    for r in records:
        punch_in = r.get("punchInTime")
        punch_out = r.get("punchOutTime")
        
        minutes = 0
        if punch_in and punch_out:
            duration = (punch_out - punch_in).total_seconds()
            minutes = int(duration / 60)
            total_seconds += duration

        writer.writerow([
            punch_in.strftime("%Y-%m-%d") if punch_in else "",
            punch_in.strftime("%H:%M") if punch_in else "",
            punch_out.strftime("%H:%M") if punch_out else "",
            minutes_to_hhmm(minutes)
        ])

    total_minutes = int(total_seconds / 60)
    writer.writerow(["TOTAL", "", "", minutes_to_hhmm(total_minutes)])

    output.seek(0)
    return Response(
        output.getvalue(),
        mimetype="text/csv",
        headers={
            "Content-Disposition": f"attachment; filename=attendance_{employee_id}_{month}.csv"
        }
    )

def minutes_to_hhmm(minutes):
    hours = minutes // 60
    mins = minutes % 60
    return f"{hours:02d}:{mins:02d}"

# -------------------- TODAY'S ATTENDANCE (FOR HR PORTAL) --------------------
@attendance_bp.route("/today", methods=["GET"])
def attendance_today():
    """
    Endpoint for HR Portal integration
    Returns today's attendance grouped by status
    """
    from datetime import datetime, time
    
    today = datetime.utcnow().date()
    today_start = datetime.combine(today, time.min)
    today_end = datetime.combine(today, time.max)
    
    # Get all attendance records for today
    records = list(
        attendance_collection.find({
            "punchInTime": {"$gte": today_start, "$lte": today_end}
        })
    )
    
    result = {
        "present": [],
        "absent": [],
        "late": [],
        "not_punched": []
    }
    
    for record in records:
        employee_id = record.get("employeeId")
        punch_in = record.get("punchInTime")
        punch_out = record.get("punchOutTime")
        
        # Determine status
        status = "present"
        if punch_in:
            # Check if late (after 9:00 AM)
            if punch_in.hour >= 9 and punch_in.minute > 0:
                status = "late"
        else:
            status = "not_punched"
        
        # Calculate hours worked
        hours_worked = None
        if punch_in and punch_out:
            duration = (punch_out - punch_in).total_seconds() / 3600
            hours_worked = round(duration, 2)
        
        # Format for HR portal
        attendance_data = {
            "id": str(record.get("_id")),
            "employee_id": employee_id,
            "employeeId": employee_id,  # Keep both for compatibility
            "employeeName": f"Employee {employee_id}",  # Placeholder - should fetch from employee service
            "department": "General",  # Placeholder - should fetch from employee service
            "date": today.isoformat(),
            "status": status,
            "checkIn": punch_in.strftime("%H:%M") if punch_in else None,
            "checkOut": punch_out.strftime("%H:%M") if punch_out else None,
            "check_in_time": punch_in.strftime("%H:%M:%S") if punch_in else None,
            "check_out_time": punch_out.strftime("%H:%M:%S") if punch_out else None,
            "hoursWorked": hours_worked,
            "location": record.get("location"),
            "wifiBSSID": record.get("wifiBSSID")
        }
        
        result[status].append(attendance_data)
    
    return jsonify(result), 200

File: backend/routes/attendanceRoutes.js
const express = require("express");
const router = express.Router();
const { db } = require("../firebaseConfig");

// Constants for Geofencing - In a real app, fetch these from DB
// Default location (e.g., Office HQ)
const DEFAULT_OFFICE = {
  lat: 12.9716,
  lng: 77.5946,
  radius: 100, // meters
  bssid: "AA:BB:CC:DD:EE:FF"
};

// Helper: Haversine Formula for distance
function getDistanceInMeters(lat1, lon1, lat2, lon2) {
  const R = 6371; // Earth radius km
  const dLat = deg2rad(lat2 - lat1);
  const dLon = deg2rad(lon2 - lon1);
  const a =
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos(deg2rad(lat1)) * Math.cos(deg2rad(lat2)) *
    Math.sin(dLon / 2) * Math.sin(dLon / 2);
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return R * c * 1000; // meters
}

function deg2rad(deg) {
  return deg * (Math.PI / 180);
}

// 1. PUNCH IN (Firestore)
router.post("/punch-in", async (req, res) => {
  try {
    const { employeeId, location, wifiBSSID } = req.body;

    // Fetch valid office config from DB (or use default)
    let office = DEFAULT_OFFICE;
    const configSnap = await db.collection('office_config').limit(1).get();
    if (!configSnap.empty) {
      office = configSnap.docs[0].data();
    }

    // Validation
    const isWifiValid = wifiBSSID === office.bssid;
    let isGeoValid = false;
    let distance = 0;

    if (location && location.latitude && location.longitude) {
      distance = getDistanceInMeters(
        location.latitude,
        location.longitude,
        office.lat || office.latitude, // handle both keys if inconsistent
        office.lng || office.longitude
      );
      isGeoValid = distance <= (office.radius || 100);
    }

    // NEW: Check if employee is late
    const now = new Date();
    const punchInTime = now.toISOString();

    // Parse office start time (e.g., "09:00")
    const officeStartTime = office.officeStartTime || "09:00";
    const [startHour, startMinute] = officeStartTime.split(':').map(Number);

    // Create today's start time
    const todayStart = new Date(now);
    todayStart.setHours(startHour, startMinute, 0, 0);

    const isLate = now > todayStart;

    // Determine Status - REQUIRE BOTH WiFi AND Geo validation
    let status = "PRESENT";
    let message = "Punch-in successful";

    if (!isWifiValid || !isGeoValid) {
      // If EITHER validation fails, mark as INVALID
      status = "INVALID_ATTENDANCE";
      if (!isWifiValid && !isGeoValid) {
        message = `Invalid WiFi and Location (${Math.round(distance)}m away)`;
      } else if (!isWifiValid) {
        message = "Invalid WiFi - Punch-in denied";
      } else {
        message = `Outside office radius (${Math.round(distance)}m away)`;
      }
    } else if (isLate) {
      // Both validations pass but late
      status = "LATE";
      message = `Punch-in successful - Late arrival (${Math.round(distance)}m from office)`;
    } else {
      // Both validations pass and on time
      status = "PRESENT";
      message = `Punch-in successful - On time (${Math.round(distance)}m from office)`;
    }

    // Create Record
    const newRecord = {
      date: now.toISOString().split('T')[0], // YYYY-MM-DD format
      employeeId,
      punchInTime,
      punchOutTime: null,
      location,
      wifiBSSID,
      status: (status === "INVALID_ATTENDANCE") ? "ABSENT" : status, // Show ABSENT immediately
      isLate, // Track if they were late
      validation: {
        wifi: isWifiValid,
        geo: isGeoValid,
        distance_meters: Math.round(distance),
        message
      }
    };

    console.log('📝 Creating attendance record for:', employeeId);
    const docRef = await db.collection('attendance').add(newRecord);
    console.log('✅ Attendance record created with ID:', docRef.id);
    console.log('   Status:', status, '| Late:', isLate, '| Message:', message);

    res.json({
      id: docRef.id,
      message,
      validation: newRecord.validation
    });

  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// 2. PUNCH OUT (Firestore) - Updated to use employeeId
router.post("/punch-out", async (req, res) => {
  try {
    const { employeeId } = req.body;

    if (!employeeId) {
      return res.status(400).json({
        error: "employeeId is required",
        message: "Please provide employeeId in request body"
      });
    }

    // Find the active attendance record (no punchOutTime yet)
    // Simplified query to avoid needing a composite index
    const snapshot = await db.collection('attendance')
      .where('employeeId', '==', employeeId)
      .where('punchOutTime', '==', null)
      .get();

    if (snapshot.empty) {
      return res.status(404).json({
        error: "No active session found",
        message: "You need to punch in first"
      });
    }

    // Sort in memory and get the most recent
    const docs = snapshot.docs.sort((a, b) => {
      const aTime = new Date(a.data().punchInTime);
      const bTime = new Date(b.data().punchInTime);
      return bTime - aTime; // desc order
    });
    const recordDoc = docs[0];
    const recordData = recordDoc.data();

    console.log('📝 Updating attendance record:', recordDoc.id, 'for employee:', employeeId);

    // Calculate hours worked
    const punchInTime = new Date(recordData.punchInTime);
    const punchOutTime = new Date();
    const hoursWorked = (punchOutTime - punchInTime) / (1000 * 60 * 60); // in hours

    // Fetch office config to get working hours
    const configSnap = await db.collection('office_config').limit(1).get();
    const office = !configSnap.empty ? configSnap.docs[0].data() : {};
    const workingHours = office.workingHours || 8;

    // Determine final status based on hours worked and initial status
    let finalStatus = recordData.status;
    let statusMessage = "";

    // CRITICAL: Check validation - if either WiFi or Geo failed, ALWAYS ABSENT
    const isValidationFailed = !recordData.validation?.wifi || !recordData.validation?.geo;

    if (isValidationFailed) {
      // Validation failed - mark as ABSENT regardless of hours worked
      finalStatus = "ABSENT";
      const failedValidations = [];
      if (!recordData.validation?.wifi) failedValidations.push("WiFi");
      if (!recordData.validation?.geo) failedValidations.push("Location");
      statusMessage = `Absent - Invalid ${failedValidations.join(" and ")} (worked ${hoursWorked.toFixed(1)}h but validation failed)`;
    } else if (recordData.status === "ABSENT") {
      // Already marked absent during punch-in, keep it
      finalStatus = "ABSENT";
      statusMessage = `Absent - ${recordData.validation?.message || "Validation failed"}`;
    } else if (hoursWorked < (workingHours * 0.5)) {
      // Worked less than 50% of working hours
      finalStatus = "ABSENT";
      statusMessage = `Absent - Only worked ${hoursWorked.toFixed(1)} hours (required ${workingHours * 0.5})`;
    } else if (Math.abs(hoursWorked - (workingHours * 0.5)) <= 0.5) {
      // Worked approximately 50% (±30 min buffer)
      finalStatus = "HALF_DAY";
      statusMessage = `Half Day - Worked ${hoursWorked.toFixed(1)} hours`;
    } else if (hoursWorked >= workingHours && recordData.isLate) {
      // Worked full hours but was late
      finalStatus = "LATE";
      statusMessage = `Present (Late) - Worked ${hoursWorked.toFixed(1)} hours`;
    } else if (hoursWorked >= workingHours) {
      // Worked full hours and on time
      finalStatus = "PRESENT";
      statusMessage = `Present - Worked ${hoursWorked.toFixed(1)} hours`;
    } else {
      // Edge case: between 50% and full hours
      finalStatus = "HALF_DAY";
      statusMessage = `Half Day - Worked ${hoursWorked.toFixed(1)} hours (required ${workingHours})`;
    }

    await recordDoc.ref.update({
      punchOutTime: punchOutTime.toISOString(),
      status: finalStatus,
      hoursWorked: parseFloat(hoursWorked.toFixed(2)),
      statusMessage
    });

    console.log('✅ Punch-out recorded for:', employeeId);
    console.log('   Hours Worked:', hoursWorked.toFixed(2), '| Final Status:', finalStatus);

    res.json({
      message: statusMessage,
      status: finalStatus,
      hoursWorked: parseFloat(hoursWorked.toFixed(2)),
      recordId: recordDoc.id
    });
  } catch (error) {
    console.error("Punch-out error:", error);
    res.status(500).json({ error: error.message });
  }
});

// 3. GET TODAY'S ATTENDANCE (Firestore)
router.get("/today", async (req, res) => {
  try {
    const todayDate = new Date().toISOString().split('T')[0]; // YYYY-MM-DD

    // Query Firestore for today's date
    const snapshot = await db.collection('attendance')
      .where('date', '==', todayDate)
      .get();

    // Fetch all employees to get names
    const employeesSnapshot = await db.collection('employees').get();
    const employeesMap = {};
    employeesSnapshot.forEach(doc => {
      const empData = doc.data();
      // Handle both emp_id and employeeId field names
      const empId = empData.employeeId || empData.employee_id || empData.emp_id;
      if (empId) {
        employeesMap[empId.toUpperCase()] = {
          name: empData.name || `${empData.firstName || ''} ${empData.lastName || ''}`.trim(),
          department: empData.department || empData.designation || 'General'
        };
      }
    });

    const records = [];

    snapshot.forEach(doc => {
      const data = doc.data();
      const empInfo = employeesMap[data.employeeId] || { name: `Employee ${data.employeeId}`, department: 'General' };

      records.push({
        id: doc.id,
        employeeId: data.employeeId,
        employeeName: empInfo.name,
        department: empInfo.department,
        date: data.date,
        punchInTime: data.punchInTime,
        punchOutTime: data.punchOutTime,
        location: data.location,
        wifiBSSID: data.wifiBSSID,
        status: data.status,
        validation: data.validation
      });
    });

    console.log(`📊 Fetched ${records.length} attendance records for ${todayDate}`);
    res.json(records);
  } catch (error) {
    console.error("Error fetching today's attendance:", error);
    res.status(500).json({ error: error.message });
  }
});

// 4. CLEAR ATTENDANCE RECORDS (Admin only)
router.delete("/clear", async (req, res) => {
  try {
    const { scope } = req.query; // 'today' or 'all'

    let snapshot;
    if (scope === 'today') {
      const todayDate = new Date().toISOString().split('T')[0];
      snapshot = await db.collection('attendance')
        .where('date', '==', todayDate)
        .get();
      console.log(`🗑️  Clearing ${snapshot.size} attendance records for today (${todayDate})`);
    } else {
      snapshot = await db.collection('attendance').get();
      console.log(`🗑️  Clearing ALL ${snapshot.size} attendance records`);
    }

    // Delete all matching documents
    const batch = db.batch();
    snapshot.docs.forEach(doc => {
      batch.delete(doc.ref);
    });
    await batch.commit();

    console.log(`✅ Successfully deleted ${snapshot.size} records`);
    res.json({
      message: `Deleted ${snapshot.size} attendance records`,
      count: snapshot.size
    });
  } catch (error) {
    console.error("Error clearing attendance records:", error);
    res.status(500).json({ error: error.message });
  }
});

module.exports = router;
